### PR TITLE
Make etcd image configurable. Add OwnerRef if object exists. Tighten functions to a private helper.

### LINKED ISSUE #9: Add OwnerRef if object already exists instead of failing
[#2/comment](https://github.com/xmudrii/etcdproxy-controller/pull/2#discussion_r192095355)

### LINKED ISSUE #10: Make etcd image property for creating ReplicaSet configurable
We're always using the `quay.io/coreos/etcd:v3.2.18` image for creating etcdproxy ReplicaSets/Pods. This option should be configurable by the controller operator.

### LINKED ISSUE #11: Tighten newReplicaSet and newService functions to a private helper
[Comment by @deads2k on #2](https://github.com/xmudrii/etcdproxy-controller/pull/2#discussion_r192134945):

> optional recommendation: I would tighten this function to a private helper that takes explicit args for its points of variability. That way any change will force compile-time errors through your code instead of silent and weird failures.

### PR DESCRIPTION
**Pull request by xmudrii**
This PR:
* Makes etcd image configurable using the `-etcd-proxy-image` flag,
* Adds OwnerRef to object (ReplicaSets and Services) if it already exists,
* Tighten functions to a private helper.

Closes #9 
Closes #10 
Closes #11

### CODE REVIEW

File: pkg/controller/etcdproxy/helpers.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+// flagfromString returns double dash prefixed flag calculated from provided key and value.
+func flagfromString(key, value string) string {
+	return fmt.Sprintf("--%s=%s", key, value)
+}
+
+// ControllerNamespace returns name of the namespace where controller is located. The namespace name is obtained
+// from the "/var/run/secrets/kubernetes.io/serviceaccount/namespace" file. In case it's not possible to obtain it
+// from that file, the function resorts to the default name, `kube-apiserver-storage`.
+func ControllerNamespace() string {
```
Comment (xmudrii) [inline pkg/controller/etcdproxy/helpers.go]: This function is going to be moved to the `cmd` package once we get to Cobra refactor (#15).

File: pkg/controller/etcdproxy/helpers.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+	if data, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace"); err == nil {
+		if ns := strings.TrimSpace(string(data)); len(ns) > 0 {
+			return ns
+		}
+	}
+
+	runtime.HandleError(fmt.Errorf("unable to detect controller namespace, resorting to default namespace name: kube-apiserver-storage"))
```
Comment (xmudrii) [inline pkg/controller/etcdproxy/helpers.go]: I believe this is a bad error handling here. Should I use `glog` to log here, or some other function?

File: pkg/controller/etcdproxy/helpers.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+	"strings"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/runtime"
+
+	samplev1alpha1 "github.com/xmudrii/etcdproxy-controller/pkg/apis/etcd/v1alpha1"
```
Comment (xmudrii) [inline pkg/controller/etcdproxy/helpers.go]: Rename the import.

File: main.go [hunk 4]
```diff
@@ -71,6 +73,8 @@ func main() {
 		URL:             coreEtcdURL,
 		CAConfigMapName: coreEtcdCAConfigMapName,
 		CertSecretName:  coreEtcdCertSecretName,
+
+		ProxyImage: proxyEtcdImage,
```
Comment (xmudrii) [inline main.go]: Potential problem here. We named the struct `coreEtcdOptions` in #2, but now the struct has the `ProxyImage` field, which is not related to the core etcd at all.

What could be a better struct name?

**Comment by sttts** [comment]
There are 3 distinct changes in this PR. Better have 3 commits.

**Comment by xmudrii** [comment]
Done.

File: main.go [hunk 5]
```diff
@@ -67,18 +70,21 @@ func main() {
 	kubeInformersNamespaced := kubeinformers.NewFilteredSharedInformerFactory(kubeClient, 10*time.Minute, controllerNamespace, nil)
 	etcdproxyInformers := informers.NewSharedInformerFactory(etcdproxyClient, 10*time.Minute)
 
-	coreEtcdOptions := &etcdproxy.CoreEtcdOptions{
-		URL:             coreEtcdURL,
-		CAConfigMapName: coreEtcdCAConfigMapName,
-		CertSecretName:  coreEtcdCertSecretName,
+	etcdProxyOptions := &etcdproxy.EtcdProxyOptions{
+		CoreEtcdOptions: etcdproxy.CoreEtcdOptions{
```
Comment (sttts) [inline main.go]: you can drop the Options postfix inside of options.

File: pkg/controller/etcdproxy/etcdproxy_controller.go [hunk 6]
```diff
@@ -69,14 +69,24 @@ const (
...
+// CoreEtcdOptions type is used to wire the core etcd information used by controller to create ReplicaSets.
 type CoreEtcdOptions struct {
 	URL             string
 	CAConfigMapName string
 	CertSecretName  string
 }
 
+// EtcdProxyOptions type is used to pass information from cli to the controller.
+type EtcdProxyOptions struct {
+	CoreEtcdOptions
```
Comment (sttts) [inline pkg/controller/etcdproxy/etcdproxy_controller.go]: No embedded please, just nesting.

**Comment by sttts** [comment]
some nits

**Comment by xmudrii** [comment]
Updated.

File: pkg/controller/etcdproxy/helpers.go [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+	}
+	replicas := int32(1)
+
+	return &appsv1.ReplicaSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      replicaSetName(etcdstorage),
+			Namespace: etcdControllerNamespace,
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(etcdstorage, schema.GroupVersionKind{
+					Group:   etcdstoragev1alpha1.SchemeGroupVersion.Group,
```
Comment (sttts) [inline pkg/controller/etcdproxy/helpers.go]: SchemeGroupVersion.WithKind

**Pull request merged**